modules/exadsBidAdapter.js

```diff
@@ -0,0 +1,511 @@
+          cid: bidData.cid,
+          width: w,
+          ttl: 360,
+          height: h,
+          netRevenue: true,
+          mediaType: mediaType,
+          meta: metaData,
+          nurl: bidData.nurl.replace(/^http:\/\//i, 'https://')
+        };
+
+        if (mediaType == 'native') {
+          bidResponse.native = native;
+        }
+
+        if (mediaType == 'video') {
+          bidResponse.vastXml = bidData.adm;
+          bidResponse.width = bidData.w;
+          bidResponse.height = bidData.h;
+        }
+
+        utils.logInfo('bidResponse->', bidResponse);
+
+        bidResponses.push(bidResponse);
+      });
+    });
+  } else {
+    imps.delete(bidRq.imp[0].id);
+    utils.logInfo('NO Ad server response ->', serverResponse.body.id);
+  }
+
+  utils.logInfo('interpretResponse -> bidResponses:', bidResponses);
+
+  return bidResponses;
+}
+
+function makeBidRequest(url, data) {
+  const payloadString = JSON.stringify(data);
+
+  return {
+    method: 'POST',
+    url: url,
+    data: payloadString
+  }
+}
+
+function getUrl(adPartner, bid) {
+  let endpointUrlMapping = {
+    [PARTNERS.ORTB_2_4]: bid.params.endpoint + '?idzone=' + bid.params.zoneId + '&fid=' + bid.params.fid
+  };
+
+  return endpointUrlMapping[adPartner] ? endpointUrlMapping[adPartner] : 'defaultEndpoint';
+}
+
+function getEnvParams() {
+  const envParams = {
+    lang: '',
+    userAgent: '',
+    osName: '',
+    page: '',
+    domain: '',
+    language: ''
+  };
+
+  envParams.domain = window.location.hostname;
+  envParams.page = window.location.protocol + '//' + window.location.host + window.location.pathname;
+  envParams.lang = navigator.language.indexOf('-') > -1
+    ? navigator.language.split('-')[0]
+    : navigator.language;
+  envParams.userAgent = navigator.userAgent;
+
+  if (envParams.userAgent.match(/Windows/i)) {
+    envParams.osName = 'Windows';
+  } else if (envParams.userAgent.match(/Mac OS|Macintosh/i)) {
+    envParams.osName = 'MacOS';
+  } else if (envParams.userAgent.match(/Unix/i)) {
+    envParams.osName = 'Unix';
+  } else if (envParams.userAgent.userAgent.match(/Android/i)) {
+    envParams.osName = 'Android';
+  } else if (envParams.userAgent.userAgent.match(/iPhone|iPad|iPod/i)) {
+    envParams.osName = 'iOS';
+  } else if (envParams.userAgent.userAgent.match(/Linux/i)) {
+    envParams.osName = 'Linux';
+  } else {
+    envParams.osName = 'Unknown';
+  }
+
+  let browserLanguage = navigator.language || navigator.userLanguage;
+  let acceptLanguage = browserLanguage.replace('_', '-');
+
+  envParams.language = acceptLanguage;
+
+  utils.logInfo('Domain -> ', envParams.domain);
+  utils.logInfo('Page -> ', envParams.page);
+  utils.logInfo('Lang -> ', envParams.lang);
+  utils.logInfo('OS -> ', envParams.osName);
+  utils.logInfo('User Agent -> ', envParams.userAgent);
+  utils.logInfo('Primary Language -> ', envParams.language);
+
+  return envParams;
+}
+
+export const imps = new Map();
```

Comment:
Why exporting a global map?

Reply:
I'm exporting it in order to make it testable. I'm using it into unit tests. An alternative is to make it as a property of the "spec" object.